### PR TITLE
Add analytics framework

### LINKED ISSUE #383: Analytics to track adoption
- How many people use it
- What plugins do they use
- What resource types do they use

Add Google Analytics for the website

Add product analytics to wash (use Bolt as an example):
- super easy to opt-out of
- checkin on startup with what subcommand is run

### PR DESCRIPTION
**Pull request by ekinanp**
Resolves https://github.com/puppetlabs/wash/issues/383

### CODE REVIEW

File: analytics/client.go
Comment (ekinanp) [inline analytics/client.go]: Will do this once everything's approved.

File: analytics/client.go
Comment (ekinanp) [inline analytics/client.go]: This is a bit hard because the version's set in `cmd.version`. Wondering if this might be worth a separate version package. Maybe properly setting the version could be a separate issue?

**Comment by ekinanp** [comment]
It'll take about a day for the data to show up in custom reports, so should probably hold off on merging until there's something we can view. However, I am seeing the data in the Realtime monitor of GA.

Note that we'll also need to create a separate view to see the usage data reported here (and also turn off "filter analytics sent from spam bots").

File: fuse/dir.go
Comment (ekinanp) [inline fuse/dir.go]: It'd be nice if there's an easy way to detect when we're auto-completing vs. invoking something. That way, we don't report analytics for auto-complete (unless we want to).

File: analytics/client.go
Comment (MikaelSmith) [reply analytics/client.go]: Reporting application version is pretty useful. A separate package that we update makes sense.

File: fuse/dir.go
Comment (MikaelSmith) [reply fuse/dir.go]: I'm not concerned about reporting for auto-complete.

File: plugin/entrySchema.go
Comment (MikaelSmith) [inline plugin/entrySchema.go]: I don't feel like this is useful enough for the added complexity. I don't agree that we need to leave out simple directories; it's easy enough to filter them out explicitly, or we can omit entries that don't have useful attributes/metadata and don't implement Read/Exec. I don't think capturing whether a plugin's types are used is something the plugin author should need to think about.

I'm also not sure the schema should be considered fragile. It's user-facing, so we shouldn't refactor it on a whim, and we won't be capturing long-running data.

File: analytics/config.go
Comment (MikaelSmith) [inline analytics/config.go]: I don't think calling Stat just before reading gives us any new information. ioutil.ReadFile should return the same error if it doesn't exist. So we should just call ReadFile.

File: api/server.go
Comment (MikaelSmith) [inline api/server.go]: I find it kind of annoying that this shows up in the activity journal:
```
Aug  6 14:37:56.528 API: POST /analytics/screenview
Aug  6 14:37:56.532 API: POST /analytics/screenview complete
```

File: cmd/internal/server/core.go
Comment (MikaelSmith) [inline cmd/internal/server/core.go]: This waits a full 250ms, right? We should stop the ticker if the goroutine completes before then.

File: cmd/root.go
Comment (MikaelSmith) [inline cmd/root.go]: This seems like it should wait for the GA flush to complete, or FlushDuration, whichever comes first. Not the full FlushDuration.

File: cmd/rootMain.go
Comment (MikaelSmith) [inline cmd/rootMain.go]: Having these run over the socket seems weird, and results in an extra journal entry for starting the Wash shell.

**Review by MikaelSmith — CHANGES_REQUESTED** [review]

File: cmd/internal/server/core.go
Comment (ekinanp) [reply cmd/internal/server/core.go]: Think it was to prevent Flush from blocking the server process from the rest of its cleanup (in case the network takes too long). I'll add a comment about that.

File: analytics/client.go
Comment (ekinanp) [reply analytics/client.go]: Should that be a separate issue or should I tack it on here? Think that may need some changes to homebrew-puppet.

File: analytics/config.go
Comment (ekinanp) [reply analytics/config.go]: The defer enforces precedence -- disabling analytics via environment has greater precedence than everything else. Otherwise, if we check environment first, then unmarshalling could overwrite our disabled value and the additional logic to check that could be annoying.

Will add a comment about this if you think it is OK.

File: cmd/rootMain.go
Comment (ekinanp) [reply cmd/rootMain.go]: Ah yeah, wanted to maintain the symmetry with the other commands. I guess we could also say "Server start" => "Someone is using Wash", so that this stuff could move to the `internal/server` package.

File: cmd/root.go
Comment (ekinanp) [reply cmd/root.go]: Yeah, makes sense. I'll add that.

File: analytics/config.go
Comment (ekinanp) [reply analytics/config.go]: Oh yes, ReadFile should be good enough. Will update.

File: plugin/entrySchema.go
Comment (ekinanp) [reply plugin/entrySchema.go]: Open to alternative ways of getting the resource type, or omitting it altogether.

I do agree that knowing what resources are most popular doesn't really have any practical meaning aside from being something "cool to know." Plugin makes sense -- if AWS is popular, we should maintain AWS and maintain it well. That automatically includes its resources. So, I think I agree with you in that we should eliminate tracking the resource type altogether and just stick to method invocations (since the latter's simple enough to track).

File: api/server.go
Comment (ekinanp) [reply api/server.go]: I don't think it's that bad. But we could ignore "/analytics" endpoint pings if it is annoying.

File: analytics/client.go
Comment (MikaelSmith) [reply analytics/client.go]: I see it as part of this issue.

File: cmd/internal/server/core.go
Comment (MikaelSmith) [reply cmd/internal/server/core.go]: Yeah, as a timer that makes sense.

File: analytics/config.go
Comment (MikaelSmith) [reply analytics/config.go]: Why not just return immediately if the environment variable is set? If it's set we don't need an ID and I don't see a reason to create the analytics file.

File: plugin/entrySchema.go
Comment (MikaelSmith) [reply plugin/entrySchema.go]: I'd be fine with that for now.

File: analytics/config.go
Comment (ekinanp) [reply analytics/config.go]: That also makes sense. Will update.

File: plugin/entrySchema.go
Comment (ekinanp) [reply plugin/entrySchema.go]: Ok. I think it's still useful for the journal to record the method invocations, so will record that stuff via typeID instead of entryType.

File: activity/core.go
Comment (ekinanp) [inline activity/core.go]: I want to keep entryType as-is here in case we do decide that reporting the entry type is useful.

**Comment by ekinanp** [comment]
I've addressed all the feedback except for the version package. Working on that next.

**Review by MikaelSmith — COMMENTED** [review]
Looks good, just waiting on adding version. .travis.yml and homebrew-puppet will need to be updated for the new path. Might make it worth adding a simple script to do release builds.

**Review by MikaelSmith — APPROVED** [review]

**Comment by MikaelSmith** [comment]
Oh, just realized we need to add a mention of this to the installation docs.

**Comment by ekinanp** [comment]
Yeah, was about to do that.

**Pull request merged**
